[rfc][pkg] check spec for module source before falling back to file in package exporter

smacke opened the pull request:
Summary: To get source for a particular module, the "correct" thing to do is to check the module's spec and use `get_source` if it's a SourceFileLoader, since subclasses may look elsewhere than the `__file__`, and the spec will give the source of truth. For torch packager, however, we prefer to use linecache, but the loader could still change the file, so we figure out the file for the module using the spec's loader rather than using `module.__file__`, if possible.

Test Plan: This code path will get exercised by CI. Also added a test for remapped files.

Differential Revision: D41412983

PaliC commented on torch/package/package_exporter.py:
```diff
@@ -422,17 +423,18 @@ def _module_exists(self, module_name: str) -> bool:
             return False
 
     def _get_source_of_module(self, module: types.ModuleType) -> Optional[str]:
+        spec = getattr(module, "__spec__", None)
```
Nit: Is it possible to do the linecache bit first, and then do this if it fails? I feel like keeping the current behavior the default and changing to this would be a more familiar user experience. Linecache itself tends to work this way. Also it would slightly improve perf as a presume linecache is faster than a direct read.

Otherwise, your changes don't seem to affect either torchdeploy/multipy or PackageImporter, so it seems reasonable.

PaliC reviewed (changes requested):
@d4l3k made a good point internally. Please add OSS unit tests to make sure things don't break.

smacke commented:
> @d4l3k made a good point internally. Please add OSS unit tests to make sure things don't break.

Yep, added this.

PaliC reviewed (approved):
Thanks for figuring out a way to simplify things by keeping the source of truth the filename.
Nit: We can go a step further by modifying the spec in Importer to have the mangled filename such that we can use spec and not worry about the second case. (This is a decent nooby task if you don't want to do it  )

smacke commented:
@pytorchbot merge -f "the one failing check appears to be unrelated with this change"

smacke committed:
[pkg] check spec for module source before falling back to file in package exporter (#90258)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/90258

To get source for a particular module, the "correct" thing to do is to check the module's spec and use `get_source` if it's a SourceFileLoader, since subclasses may look elsewhere than the `__file__`, and the spec will give the source of truth. For torch packager, however, we prefer to use linecache, but the loader could still change the file, so we figure out the file for the module using the spec's loader rather than using `module.__file__`, if possible.

Test Plan: This code path will get exercised by CI. Also added a test for remapped files.

Reviewed By: PaliC

Differential Revision: D41412983

fbshipit-source-id: 435ebeef0c13f86c73ae1f0b7cc2c8d6793793a4

smacke commented:
@pytorchbot merge -f "this change previously ran OK on CI and doesn't have anything new"
